### ISSUE TITLE
Time-series sensor ID should be a route parameter

### ISSUE THREAD

**Issue by yousefamar** [me-box/databox-store-blob#15]
For the [time-series API](https://github.com/me-box/databox-store-blob/blob/macaroons/README.md#time-series-data-note-this-may-move-to-ts-after-mozfest) (if that's even here and not its own repo?) we should have sensor_id be a [route parameter](http://expressjs.com/en/guide/routing.html#route-parameters) so that path permissions cover individual sensors for any given driver-store pair. E.g. `/api/data/:sensorID/latest`, and that way some apps could only have access to `[ "/api/data/accelerometer/latest", "/api/data/light/*" ]` while another has `"/api/data/*"` for example to replicate original behaviour.

**Comment by Toshbrown** [comment]
Yes, I think it should. I need to chat to Tom before making the changes, though. 

moving to /api/data/:sensorID/latest will al help with internal logging.

**Comment by yousefamar** [comment]
Cool, so `databox-store-blob` will be a hybrid of time-series data and JSON documents? And in that case should it be renamed to `databox-store-general` or so instead of `databox-store-json`? I can think of pros and cons for either way, but am personally indifferent.

In that case, especially to make it so that #17 is a bit more consistent, could we refactor to something like `/:api/:type/:query`, e.g.:

- /(data|actuate|sub|unsub)
  - /json
    - /:key
  - /ts
    - /:sensor
- /logs
  - [log routes]
  - ...

The exceptions are `/cat` (POST for drivers to add, GET for apps to query catalogue), and `/status`, both at root. Also I think it might be clearer for permissions if we serve WS at something like `/ws` (controlled with the `path` option in WebSocket server).

So for example, here's the path caveat for some app:
```
[
  "/cat",                        // Can access (potentially censored or filtered) second-level store catalogue
  "/ws",                         // Can connect to store WS server
  "/data/json/profile",          // Can access this specific JSON document
  "/data/ts/accelerometer/*",    // Can access all endpoints for accelerometer (root, latest, since, range)
  "/data/ts/gps/latest",         // Can only see latest GPS reading
  "/logs/*",                     // Can read any logs (for example)
  "/(data|sub|unsub)/ts/light/*" // Can read any light data, and also subscribe or unsubscribe to it
]
```

P.S. Also maybe instead of `data` and `actuate`, I feel like it would be more elegant to abstract those into `read` and `write`. That way, you have four possibilities, all controlled by path caveats:

- An app can `/read` data (or `/sub` to a data stream) then process it
- A driver can `/write` data to `*` paths
- An app can `/write` data to certain paths
- A driver can `/read` data (or `/sub` to a data stream) on certain paths, and act on it (actuate)

Thoughts? Should I go for it in #17?

**Comment by Toshbrown** [comment]
Yep, that makes a lot of sense. The move to read and write is very sensible :+1: 

I'm still in two minds about splitting this into two stores. But with your schema above it would not be difficult to do after this change so let's go with your plan.

**Comment by yousefamar** [comment]
Gonna take a stab at this and #17 now. Do you think it would be confusing if we started merging things to `master` on their respective repos? For your demo we could tag the current masters `0.1-alpha` or similar following [semver](http://semver.org/), or else branch `alpha` at current master and leave it there.

**Comment by Toshbrown** [comment]
Let's leave master for now. Branch from hypercat then merge back into hypercat I can then do a big test before merging all the  hypercat branches back into master for all the components. 

Unless that's going to cause you any problems?

**Comment by yousefamar** [comment]
Not at all, will do. :+1:

**Comment by yousefamar** [comment]
One thing I noticed for #20 was that for most of the REST endpoints, GET and POST implicitly meant read and write respectively. My thoughts on that were that the store should be indifferent to what HTTP method is used, and only care about the path (i.e. `/read` and `/write` — I made some comments where I think `.get` and `.post` should be `.all` with the `req.body` and `req.query` shim).

I think this would make it easier for app devs who might use weird tools/frameworks to interface with the API, and align more with how people (ab)use HTTP methods today interchanging GET query strings and POST bodies to do the same stuff, neither of which were what they were meant for when that was all first defined. As a bonus, this makes caveats simpler since we already have paths, so no need to add extra stuff to differentiate between GET /some-path and POST /some-path.
